networkx/convert_matrix.py

```diff
@@ -1170,83 +1171,38 @@ def to_numpy_array(
 
     if nodelist is None:
         nodelist = list(G)
-        nodeset = G
-        nlen = len(G)
-    else:
-        nlen = len(nodelist)
-        nodeset = set(G.nbunch_iter(nodelist))
-        if nlen != len(nodeset):
-            for n in nodelist:
-                if n not in G:
-                    raise nx.NetworkXError(f"Node {n} in nodelist is not in G")
-            raise nx.NetworkXError("nodelist contains duplicates.")
-
-    undirected = not G.is_directed()
-    index = dict(zip(nodelist, range(nlen)))
+    nlen = len(nodelist)
+
+    # Input validation
+    nodeset = set(nodelist)
+    if nodeset - set(G):
+        raise nx.NetworkXError(f"Nodes {nodeset - set(G)} in nodelist is not in G")
+    if len(nodeset) < nlen:
+        raise nx.NetworkXError("nodelist contains duplicates.")
+
+    A = np.full((nlen, nlen), fill_value=nonedge, dtype=dtype, order=order)
+
+    # Corner cases: empty nodelist or graph without any edges
+    if nlen == 0 or G.number_of_edges() == 0:
+        return A
+
+    # Map nodes to row/col in matrix
+    idx = dict(zip(nodelist, range(nlen)))
+    G = G.subgraph(nodelist)
```

Comment:
Does this make it faster?  The subgraph view is slow if you access many edges... especially if you do them more than once each. I believe copying the subgraph to a graph speeds it up.  But it might be better just to check each edge like the old code does (explicit set checking for MultiGraph and looking for a keyerror with Graph).
```suggestion
    if len(nodelist) < len(G):
        G = G.subgraph(nodelist).copy()
```